Tutorials/Tutorial27_PostProcessing/src/Tutorial27_PostProcessing.cpp

```diff
@@ -0,0 +1,615 @@
+        const auto PS = CreateShader(m_pDevice, nullptr, "ApplyToneMap.fx", "ApplyToneMapPS", SHADER_TYPE_PIXEL, Macros);
+
+        PipelineResourceLayoutDescX ResourceLayout;
+        ResourceLayout
+            .AddVariable(SHADER_TYPE_PIXEL, "cbToneMappingAttibs", SHADER_RESOURCE_VARIABLE_TYPE_STATIC)
+            .AddVariable(SHADER_TYPE_PIXEL, "g_TextureHDR", SHADER_RESOURCE_VARIABLE_TYPE_DYNAMIC);
+
+        RenderTech.InitializePSO(m_pDevice,
+                                 nullptr, "Tutorial27_PostProcessing::ApplyToneMap",
+                                 VS, PS, ResourceLayout,
+                                 {m_pSwapChain->GetDesc().ColorBufferFormat},
+                                 TEX_FORMAT_UNKNOWN,
+                                 DSS_DisableDepth, BS_Default, false);
+        ShaderResourceVariableX{RenderTech.PSO, SHADER_TYPE_PIXEL, "cbToneMappingAttibs"}.Set(m_Resources[RESOURCE_IDENTIFIER_TONE_MAPPING_CONSTANT_BUFFER].AsBuffer());
+        RenderTech.InitializeSRB(true);
+    }
+
+    const Uint32 CurrFrameIdx = (m_CurrentFrameNumber + 0x0) & 0x1;
+
+    ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "g_TextureHDR"}.Set(m_Resources[RESOURCE_IDENTIFIER_RADIANCE0 + CurrFrameIdx].GetTextureSRV());
+
+    ScopedDebugGroup DebugGroup{m_pImmediateContext, "ApplyToneMap"};
+
+    ITextureView* pRTV = m_pSwapChain->GetCurrentBackBufferRTV();
+    m_pImmediateContext->SetRenderTargets(1, &pRTV, nullptr, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
+    m_pImmediateContext->SetPipelineState(RenderTech.PSO);
+    m_pImmediateContext->CommitShaderResources(RenderTech.SRB, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
+    m_pImmediateContext->Draw({3, DRAW_FLAG_VERIFY_ALL});
+    m_pImmediateContext->SetRenderTargets(0, nullptr, nullptr, RESOURCE_STATE_TRANSITION_MODE_NONE);
+}
+
+void Tutorial27_PostProcessing::UpdateUI()
+{
+    ImGui::PushFont(static_cast<ImFont*>(m_pImGuiFont));
+    ImGui::SetNextWindowPos(ImVec2(10, 10), ImGuiCond_FirstUseEver);
+    if (ImGui::Begin("Settings", nullptr, ImGuiWindowFlags_AlwaysUseWindowPadding | ImGuiWindowFlags_AlwaysAutoResize))
+    {
+        ImGui::SetNextItemOpen(true, ImGuiCond_FirstUseEver);
+
+        if (ImGui::TreeNode("Rendering"))
+        {
+            // clang-format off
+            ImGui::SliderFloat("IBL Scale", &m_ShaderParams->IBLScale, 0.01f, 4.0f);
+            ImGui::SliderFloat("Ambient Occlusion Scale", &m_ShaderParams->OcclusionScale, 0.0f, 1.0f);
+            ImGui::SliderFloat("Screen Space Reflection Scale", &m_ShaderParams->ReflectionScale, 0.0f, 1.0f);
+            ImGui::Checkbox("Enable Animation", &m_IsAnimationActive);
+            // clang-format on
+            ImGui::TreePop();
+        }
+
+        ImGui::SetNextItemOpen(true, ImGuiCond_FirstUseEver);
+        if (ImGui::TreeNode("Post Processing"))
+        {
+            m_ScreenSpaceReflection->UpdateUI(m_ShaderParams->SSRSettings);
```

Comment:
I removed the tree node from SSR, so it needs to be added here.

Reply:
Fixed